libcudacxx/include/cuda/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,431 @@
+                       "Requested IPC memory handle type not supported");
+
+    ::cudaMemPoolProps __pool_properties{};
+    __pool_properties.allocType     = ::cudaMemAllocationTypePinned;
+    __pool_properties.handleTypes   = ::cudaMemAllocationHandleType(__allocation_handle_type);
+    __pool_properties.location.type = ::cudaMemLocationTypeDevice;
+    __pool_properties.location.id   = __device_id;
+    ::cudaMemPool_t __cuda_pool_handle{};
+    _CCCL_TRY_CUDA_API(::cudaMemPoolCreate, "Failed to call cudaMemPoolCreate", &__cuda_pool_handle, &__pool_properties);
+
+    // CUDA drivers before 11.5 have known incompatibilities with the async allocator.
+    // We'll disable `cudaMemPoolReuseAllowOpportunistic` if cuda driver < 11.5.
+    // See https://github.com/NVIDIA/spark-rapids/issues/4710.
+    int __driver_version = 0;
+    _CCCL_TRY_CUDA_API(::cudaDriverGetVersion, "Failed to call cudaDriverGetVersion", &__driver_version);
+
+    constexpr int __min_async_version = 11050;
+    if (__driver_version < __min_async_version)
+    {
+      int __disable_reuse = 0;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemPoolSetAttribute,
+        "Failed to call cudaMemPoolGetAttribute",
+        __cuda_pool_handle,
+        ::cudaMemPoolReuseAllowOpportunistic,
+        &__disable_reuse);
+    }
+
+    _CCCL_TRY_CUDA_API(
+      ::cudaMemPoolSetAttribute,
+      "Failed to call cudaMemPoolReuseAllowOpportunistic",
+      __cuda_pool_handle,
+      ::cudaMemPoolAttrReleaseThreshold,
+      &__release_threshold);
+
+    // allocate the requested initial size to pprime the pool
+    if (__initial_pool_size != 0)
+    {
+      void* __ptr{nullptr};
+      _CCCL_TRY_CUDA_API(
+        ::cudaMallocAsync,
+        "cuda_memory_pool failed allocate the initial pool size",
+        &__ptr,
+        __initial_pool_size,
+        ::cudaStream_t{0});
+
+      _CCCL_ASSERT_CUDA_API(
+        ::cudaFreeAsync, "cuda_memory_pool failed to free the initial pool allocation", __ptr, ::cudaStream_t{0});
+    }
+    return __cuda_pool_handle;
+  }
+
+public:
+  /**
+   * @brief Constructs a cuda_memory_pool with the optionally specified initial pool size
+   * and release threshold.
+   *
+   * If the pool size grows beyond the release threshold, unused memory held by the pool will be
+   * released at the next synchronization event.
+   *
+   * @throws ::cuda::cuda_error if the CUDA version does not support `cudaMallocAsync`
+   *
+   * @param initial_pool_size Optional initial size in bytes of the pool. If no value is provided,
+   * initial pool size is half of the available GPU memory.
+   * @param release_threshold Optional release threshold size in bytes of the pool. If no value is
+   * provided, the release threshold is set to the total amount of memory on the current device.
+   */
+  cuda_memory_pool(
+    const size_t initial_pool_size                             = 0,
```

Comment:
Should these be `std::optional` rather than defaulted?

Reply:
I do not see that much of a benefit of making this `optional`

Its a hefty compile time cost for only a little clarity

Reply:
OK. I thought it was the modern way to do optional parameters in C++. :)

Reply:
I am not sure about function arguments with a sensible default. Its more usefull as a return type of a function